### DISCUSSION TITLE
开源之夏 2021 | 讨论 | Apollo 公共 namespace 支持 xml、json、yaml 等格式

### DISCUSSION THREAD

**Discussion by dake0805** [apolloconfig/apollo#3815] [Show and tell]
[Summer 2021 of Open Source Promotion Plan](https://summer.iscas.ac.cn/#/org/orgdetail/apollo?lang=chi)

目标：使Apollo 公共 namespace 支持 xml、json、yaml 等格式，解决 [Issue(#2602)](https://github.com/ctripcorp/apollo/issues/2602)

需求拆分：

- 能够支持创建公共的 xml json yaml 等非 properties 格式 namespace，以及 Java 客户端能够以类似获取公共的 properties 格式的 namespace 的方法获取这些公共 namespace。
- 实现继承功能，因非 properties 格式本身不支持键值粒度的操作，继承将在整个文件内容粒度进行。
- 支持 yaml 格式的公共 namespace 的键值级别解析，类似当前私有 namespace 中的 yaml 格式。

这三点需求为递进关系，后续将根据进度安排、功能讨论等情况，根据这三点需求实际进行开发。

后续相关进度情况、出现的问题及后续讨论将于此 discussion 贴出和讨论。

**Comment by lepdou** [comment]
@dake0805 是参与开源之夏的同学，非常感谢对 Apollo 开源的支持~

针对这次项目本身，我和 @dake0805 一起讨论了一下需求，拆分成三个层次的目标，我个人倾向于实现前面两层就够了。也就是

- 能够支持创建公共的 xml json yaml 等非 properties 格式 namespace，以及 Java 客户端能够以类似获取公共的 properties 格式的 namespace 的方法获取这些公共 namespace。

- 实现继承功能，因非 properties 格式本身不支持键值粒度的操作，继承将在整个文件内容粒度进行。

其它同学帮忙一起看一下呢~

**Comment by zouyx** [comment]
我有个问题：在公私有的情况下，最主要需要解决的就是 key 覆盖的问题，xml、json 的继承将在整个文件内容粒度进行，是什么意思？例如：
父 json
```json
{
"A":"A",
“B”:“C"
}
```

子json
```json
{
"B":"B"
}
```

那客户端得到的是？

**Reply by lepdou** [reply]
得到的是这个：
{
"B":"B"
}

**Reply by zouyx** [reply]
ok

**Comment by nobodyiam** [comment]
> 能够支持创建公共的 xml json yaml 等非 properties 格式 namespace，以及 Java 客户端能够以类似获取公共的 properties 格式的 namespace 的方法获取这些公共 namespace。
实现继承功能，因非 properties 格式本身不支持键值粒度的操作，继承将在整个文件内容粒度进行。

第一和第二点比较清楚，可以着手设计和开发了，不过有一点需要注意的是我们需要在 UI 和文档中明确得让用户感知到 `非 properties 格式的 namespace 继承将在整个文件内容粒度进行`这个点。

> 支持 yaml 格式的公共 namespace 的键值级别解析，类似当前私有 namespace 中的 yaml 格式。

第三点作为高级特性可以继续讨论下必要性和实现成本，之前也确实有社区同学提了相关的需求(#2813)。

**Reply by lepdou** [reply]
> 非 properties 格式的 namespace 继承将在整个文件内容粒度进行这个点.

这个考虑了，需要再 UI 层面能够体现出来，另外也搞一个类似于代码比对效果，左右两侧对比显示文本内容

**Comment by lepdou** [comment]
@dake0805 需求层面就这么定了吧，我们就先按照前面两个目标去实现。

**Reply by tianmingxing** [reply]
有实施计划了吗？

**Reply by dake0805** [reply]
@tianmingxing see [release/v1.9.0](https://github.com/apolloconfig/apollo/releases/tag/v1.9.0)
